openff/bespokefit/bespoke/smirks.py

```diff
@@ -158,148 +180,232 @@ def _get_all_bespoke_smirks(
         """
         bond_smirks = []
-        for bond in molecule.bonds:
-            atoms = (bond.atom1_index, bond.atom2_index)
+        fragment_mol = fragment_data.fragment_molecule
+        # group the bonds
+        bonds = [(bond.atom1_index, bond.atom2_index) for bond in fragment_mol.bonds]
+        bond_groups = SmirksGenerator.group_valence_by_symmetry(fragment_mol, bonds)
+        bond_groups = [*bond_groups.values()]
+        for bond_group in bond_groups:
+            # make a list of the target molecules and bonds, only add the parent if it is different
+            target_atoms = [bond_group]
+            target_molecules = [fragment_mol]
+            if not fragment_is_parent:
+                # map the bonds to the parent and use both in the cluster graph
+                parent_bonds = [
+                    tuple(fragment_data.fragment_parent_mapping[i] for i in bond)
+                    for bond in bond_group
+                ]
+                target_atoms.append(parent_bonds)
+                target_molecules.append(fragment_data.parent_molecule)
+
+            smirks = self._get_new_cluster_graph_smirks(
+                atoms=target_atoms, molecules=target_molecules
+            )
             new_smirks = BespokeBondSmirks(
-                smirks=self._get_new_single_graph_smirks(
-                    atoms=atoms, molecule=molecule
-                ),
-                k=0,
-                length=0,
-                atoms=[
-                    atoms,
-                ],
+                smirks=smirks, k=0, length=0, atoms=bond_group
             )
             if new_smirks not in bond_smirks:
                 bond_smirks.append(new_smirks)
             else:
                 # update the covered atoms
                 index = bond_smirks.index(new_smirks)
-                bond_smirks[index].atoms.add(atoms)
+                bond_smirks[index].atoms.add(bond_group)
 
         return bond_smirks
 
-    def _get_bespoke_angle_smirks(self, molecule: Molecule) -> List[BespokeAngleSmirks]:
+    def _get_bespoke_angle_smirks(
+        self, fragment_data: FragmentData, fragment_is_parent: bool
+    ) -> List[BespokeAngleSmirks]:
         """
         For the molecule generate a unique set of bespoke angle smirks.
         """
         angle_smirks = []
-        for angle in molecule.angles:
-            atom_ids = tuple([atom.molecule_atom_index for atom in angle])
+        fragment_mol = fragment_data.fragment_molecule
+        angles = [
+            tuple([atom.molecule_atom_index for atom in angle])
+            for angle in fragment_mol.angles
+        ]
+        angle_groups = SmirksGenerator.group_valence_by_symmetry(fragment_mol, angles)
+        angle_groups = [*angle_groups.values()]
+
+        for angle_group in angle_groups:
+            target_atoms = [angle_group]
+            target_molecules = [fragment_data.parent_molecule]
+            if not fragment_is_parent:
+                parent_angles = [
```

Comment:
Could you add some tests to make sure the logic in the `if not fragment_is_parent:` conditionals is tested?

Reply:
Great shout, this also uncovered a few bugs caused by typos so thanks for this!